src/libre/ast_new_from_fsm.c

```diff
@@ -0,0 +1,474 @@
+				return 0;
+			}
+			if (!ast_add_expr_concat(cat, tmp)) {
+				ast_expr_free(tmp);
+				ast_expr_free(cat);
+				return 0;
+			}
+
+			if (!e->expr) {
+				e->expr = cat;
+			} else {
+				if (!change_to_alt(&e->expr)) {
+					ast_expr_free(cat);
+					return 0;
+				}
+
+				if (!ast_add_expr_alt(e->expr, cat)) {
+					ast_expr_free(cat);
+					return 0;
+				}
+			}
+		}
+	}
+
+	free_restate(state);
+}
+
+struct ast *
+ast_new_from_fsm(const struct fsm *fsm)
+{
+	struct ast *ast;
+	struct ast_expr *expr;
+	struct fsm_options opt;
+
+	unsigned int numstates;
+	fsm_state_t start;
+	struct rese *restates;
+	struct rese *restart, *reend;
+
+	unsigned int i;
+	struct rese *state;
+
+	numstates = fsm_countstates(fsm);
+
+	if (!fsm_getstart(fsm, &start))
+		return 0;
+
+	/* Copy FSM states as RE states.
+	 * Reserve an additional state to use as an artificial single
+	 * end state. */
+	restates = build_restates(numstates + 1);
+	if (!restates)
+		return 0;
+
+	restart = &restates[start];
+	reend = &restates[numstates];
+
+	/* Copy FSM edges as RE edges */
+	if (!fsm_walk_edges(fsm, restates, build_edge_literal, build_edge_epsilon)) {
+		free_restates(restates, numstates + 1);
+		return 0;
+	}
+
+	/* Add epsilon transitions from all FSM end states to our
+	 * artificial end state. */
+	for (i = 0; i < numstates; i++) {
+		if (!fsm_isend(fsm, i))
+			continue;
+
+		if (!build_edge_epsilon(fsm, i, numstates, restates)) {
+			free_restates(restates, numstates + 1);
+			return 0;
+		}
+	}
+
+	/* Remove all the states that are not the start or end state */
+	for (i = 0; i < numstates; i++) {
+		if (i == start)
+			continue;
+
+		if (!remove_state(&restates[i])) {
+			free_restates(restates, numstates + 1);
+			return 0;
+		}
+	}
+
+	/* Finally we are left with the start state, end state, a transition
+	 * from start to end, and optionally also a transition from start to
+	 * start. If the latter is present, handle it here. */
+	expr = restart->succnext->expr;
+
+	if (restart->expr) {
+		struct ast_expr *cat = ast_make_expr_concat(0);
+		if (!cat) {
+			free_restates(restates, numstates + 1);
+			return NULL;
+		}
+
+		struct ast_expr *rep = ast_make_expr_repeat(
+			0, restart->expr, ast_make_count(0, NULL, AST_COUNT_UNBOUNDED, NULL));
+		if (!rep) {
+			ast_expr_free(cat);
+			free_restates(restates, numstates + 1);
+			return NULL;
+		}
+
+		if (!ast_add_expr_concat(cat, rep)) {
+			ast_expr_free(rep);
+			ast_expr_free(cat);
+			free_restates(restates, numstates + 1);
+			return NULL;
+		}
+
+		if (!ast_add_expr_concat(cat, expr)) {
+			ast_expr_free(cat);
+			free_restates(restates, numstates + 1);
+			return NULL;
+		}
+
+		expr = cat;
+	}
+
+	restart->succnext->expr = NULL;
+	restart->expr = NULL;
+
+	free_restates(restates, numstates + 1);
+
+	ast = ast_new();
+	if (!ast) {
+		ast_expr_free(expr);
+		return NULL;
+	}
+
+	ast->expr = expr;
+
+	if (!ast_rewrite(ast, 0)) {
+		ast_free(ast);
+		return NULL;
+	}
+
+	return ast;
```

Comment:
Testing this function with some random FSMs (from `fuzz/gengraph $RANDOM 10 5 5 abcd 0`) I found it's very easy to get `ast_new_from_fsm` to fail. That's fine, if it's for a reason that makes sense, but I can't work out what it is! Am I missing something? I tried with DFA, too.

Can you set `errno` for the cases where we return failure here please? `EINVAL` especially. Currently we indicate failure as "ast_new_from_fsm: Success", which should never be visible.

Reply:
We should only return failure if another function failed, in which case that function should have set `errno`. I see now that I'm returning `NULL` for FSMs without a start state. I shouldn't be doing that, I should be returning `AST_EXPR_TOMBSTONE`: it's a valid FSM but can never match anything. There may be more cases like that, I'll check!

Reply:
Fixed the ones I saw, and then the fuzzer found something that fails in `ast_rewrite` but is also reproducible from a straight PCRE regex, will look into that and put that in a separate PR if I find anything!

Reply:
Brilliant!

Reply:
I can still produce `ast_new_from_fsm: Success` for this entertaining-looking FSM:
![image](https://user-images.githubusercontent.com/1371085/93689353-bf46c480-fa82-11ea-811c-f8a9a0b363ea.png)

I suspected because no end state is reachable. But this works:
```
; echo 'start: 0;' | ./build/bin/re -pl pcre -f -
```

I trimmed down a test case, and can reproduce it with this:
![image](https://user-images.githubusercontent.com/1371085/93689460-8e1ac400-fa83-11ea-98b8-1f089087c924.png)

```
0 -> 1 'a';

start: 2; # same error if this is 1 instead
end: 0;
```

Reply:
Huh, that is surprising. That exact test case does work here:
```
$ cat test.fsm
0 -> 1 'a';

start: 2; # same error if this is 1 instead
end: 0;
$ build/bin/re -f -pl pcre test.fsm
[^\x00-\xff]
$
```

Reply:
After a rebuild with optimizations enabled, I do get the result you are seeing. Will try to figure out what is going on.

Reply:
Ah, the risks of compiling without warnings! There was a missing `return` statement in `remove_state`. In non-optimized builds, I happened to get a non-zero return value that indicates success, in optimized builds I (and you) happened to get a zero return value that indicates an error. That is fixed now.

Reply:
Also visible for this case:
![image](https://user-images.githubusercontent.com/1371085/93690553-6ed56400-fa8e-11ea-928a-a061f5e95d95.png)

Reply:
Yeah, it would have happened for many cases. To clarify: are you still seeing this now? I pushed a fix for this, and on the earlier commit, I was able to get that same (non-)error for `a?b` (only with optimizations enabled), on the latest it works, with or without optimizations.

Looking more closely at the output of the compiler, there was also one place where I was using a C99 feature, a declaration after a statement. I will push a fix for that too.

Reply:
You know you can compile with warnings too! `bmake -r CC=gcc` should do it.

Reply:
Confirmed that all these case work for me now, thank you!